File: task2.scala
import java.io.{File, PrintWriter}

import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import org.apache.spark.storage.StorageLevel

import scala.collection.mutable._

object task2 {
  def main(args: Array[String]): Unit = {
    val t1 = System.nanoTime
    Logger.getLogger("org").setLevel(Level.ERROR)

    val ss = SparkSession
      .builder()
      .appName("second")
      .config("spark.master", "local[*]")
      .getOrCreate()
    val sc = ss.sparkContext

    val filter_threshold=args(0).toInt
    val threshold=args(1).toInt
    val input_path=args(2)
    val output_path=args(3)

    type pairRDD=(String,List[String])
    def edit_data(in:Array[String]): pairRDD ={
      var temp=""
      temp=temp.concat(in(0).slice(1,in(0).length-1))
      temp=temp.concat("-")
      temp=temp.concat(in(1).slice(1,in(1).length-1))
      var temp1=""
      temp1=temp1.concat(in(5).slice(1,in(5).length-1).replaceFirst("^0+(?!$)", ""))
      return (temp,List(temp1))
    }
    val largeRDD = sc.textFile(input_path).map(line=>line.split(',')).filter(line=> line(0)!="TRANSACTION_DT").map(edit_data)
    val basketsRDD = largeRDD.reduceByKey((a,b)=>a++b).filter(line=>line._2.size>filter_threshold).persist(StorageLevel.DISK_ONLY)
    for(x<-basketsRDD.take(10)){
      println(x)
    }
    type pair=(String,Int)
    val total_baskets=basketsRDD.count()

    //defining apriori function
    def apriori(iterator: Iterator[pairRDD]): Iterator[pair] = {
      var baskets=new ListBuffer[List[String]]()
      var baskets_in_partition=0
      var itemset_size=1
      var candidate:List[pair]=List()
      var temp_candidate=new ListBuffer[pair]()
      var frequent:Set[String]=Set()
      var frequent_itemset:List[String]=List()
      //calculating number of baskets in the partition
      //creating a list of baskets from the iterator
      //adding all possible singletons in frequent set
      for (elem <- iterator) {
        var temp=new ListBuffer[String]()
        baskets_in_partition+=1
        for(item<-elem._2) {
          temp+=item
          frequent+=item
        }
        var temp_basket=temp.toList
        baskets+=temp_basket
      }
      var list_of_baskets=baskets.toList
      //creating initial frequent itemsets of size1
      var temp_itemset=new ListBuffer[String]()
      for(item<-frequent){
        temp_itemset+=item
      }
      frequent_itemset=temp_itemset.toList
      //calculating reduced threshold for the partition
      var reduced_threshold=((baskets_in_partition*1.0/total_baskets)*threshold*1.0)
      //start a while loop for getting all the candidates
      while(frequent_itemset.size!=0){
        //println(frequent_itemset)
        //counting the frequency of each itemset in the list of baskets
        val frequency=Map[String,Int]()
        //for each basket in list of baskets
        for(basket<-list_of_baskets){
          val temp_basket=basket.toSet
          //for each tuple in frequent itemsets
          for(tpl<-frequent_itemset){
            //check if all the elements of the tuple are present in the basket
            var flag=1
            var temp_tpl=tpl.split(",")
            for(item<-temp_tpl){
              if(!temp_basket.contains(item)){
                flag=0
              }
            }
            //if all are present, increase the frequency of that tuple by 1
            if(flag==1){
              if(frequency.contains(tpl)){
                frequency(tpl)+=1
              }
              else {
                frequency(tpl)=1
              }
            }
          }
        }
        //clear the frequent itemset to put actually frequent itemsets
        temp_itemset.clear()
        frequent_itemset=temp_itemset.toList
        //check if frequency is greater than reduced threshold
        for((k,v)<-frequency){
          if(v>=reduced_threshold){
            temp_candidate+=((k,1))
            temp_itemset+=k
          }
        }
        //add actual frequents to the frequent itemsets
        frequent_itemset=temp_itemset.toList
        //println("actual frequent itemsets are:"+frequent_itemset)
        temp_itemset.clear()
        //flatten the tuple into a list of items
        for(tpl<-frequent_itemset){
          var temp_tpl=tpl.split(",")
          for(item<-temp_tpl){
            temp_itemset+=item
          }
        }
        //get the list of frequent items from the buffer after flattening
        frequent_itemset=temp_itemset.toList
        temp_itemset.clear()
        itemset_size+=1
        //convert the list to set and get all possible subsets
        var all_subsets=frequent_itemset.toSet[String].subsets.map(_.toList).toList
        //get only subsets of size= itemset_size
        for(subset<-all_subsets){
          if(subset.size==itemset_size){
            var temp_subset=subset.sorted
            var temp=""
            for(i<-0 to (temp_subset.size-2)){
              temp = temp.concat(temp_subset(i))
              temp = temp.concat(",")
            }
            temp = temp.concat(temp_subset(temp_subset.size-1))
            temp_itemset+=temp
          }
        }
        frequent_itemset=temp_itemset.toList
        //println(frequency)
      }
      candidate=temp_candidate.toList
      //println(candidate)

      return candidate.toIterator
    }

    val candidates=basketsRDD.mapPartitions(apriori).reduceByKey((a,b)=> 1).map(line=> line._1).collect()
    var sorted_candidates=candidates.sorted
    //write the output to the file in proper order you will have to process according to the size
    //for(x<-sorted_candidates){
    //  println(x)
    //}
    //println("*************************************************")
    //defining function for finding the actual frequent pairs
    def apriori_1(iterator: Iterator[pairRDD]): Iterator[pair] = {
      var baskets=new ListBuffer[List[String]]()
      //creating a list of baskets from the iterator
      for (elem <- iterator) {
        var temp=new ListBuffer[String]()
        for(item<-elem._2) {
          temp+=item
        }
        var temp_basket=temp.toList
        baskets+=temp_basket
      }
      var list_of_baskets=baskets.toList
      //counting the frequency of each itemset in the list of baskets
      val frequency=Map[String,Int]()
      //for each basket in list of baskets
      for(basket<-list_of_baskets){
        var temp_basket=basket.toSet
        //for each tuple in candidates
        for(tpl<-candidates){
          //check if all the elements of the tuple are present in the basket
          var flag=1
          var temp_tpl=tpl.split(",")
          for(item<-temp_tpl){
            if(!temp_basket.contains(item)){
              flag=0
            }
          }
          //if all are present, increase the frequency of that tuple by 1
          if(flag==1){
            if(frequency.contains(tpl)){
              frequency(tpl)+=1
            }
            else {
              frequency(tpl)=1
            }
          }
        }
      }
      var frequent_itemset:List[pair]=List()
      var temp_frequent_itemset=new ListBuffer[pair]()
      //check if frequency is greater than reduced threshold
      for((k,v)<-frequency){
        if(v>=threshold){
          temp_frequent_itemset+=((k,v))
        }
      }
      frequent_itemset=temp_frequent_itemset.toList
      //println(frequent_itemset)
      //println("*************************************************")
      return frequent_itemset.toIterator
    }

    var frequent_itemsets=basketsRDD.mapPartitions(apriori_1).reduceByKey((a,b)=>a+b).filter(line=> line._2>=threshold).map(line=>line._1).collect()
    var sorted_frequent_itemsets=frequent_itemsets.sorted
    // writing output to a file
    val out_file = new File(output_path)
    val print_Writer = new PrintWriter(out_file)
    print_Writer.write("Candidates:\n")
    var size=1
    var flag=1
    while(flag!=0){
      flag=0
      for(x<-sorted_candidates){
        var temp=x.split(",")
        if(temp.size==size){
          if(flag==0 && size>=2){
            print_Writer.write("\n\n(")
          }
          else if(flag==0){
            print_Writer.write("(")
          }
          else{
            print_Writer.write(",(")
          }
          flag=1
          for(i<-0 to (temp.size-2)){
            print_Writer.write("'"+temp(i)+"'"+", ")
          }
          print_Writer.write("'"+temp(temp.size-1)+"'"+")")
        }
      }
      size+=1
    }

    print_Writer.write("\n\nFrequent Itemsets:\n")
    size=1
    flag=1
    while(flag!=0){
      flag=0
      for(x<-sorted_frequent_itemsets){
        var temp=x.split(",")
        if(temp.size==size){
          if(flag==0 && size>=2){
            print_Writer.write("\n\n(")
          }
          else if(flag==0){
            print_Writer.write("(")
          }
          else{
            print_Writer.write(",(")
          }
          flag=1
          for(i<-0 to (temp.size-2)){
            print_Writer.write("'"+temp(i)+"'"+", ")
          }
          print_Writer.write("'"+temp(temp.size-1)+"'"+")")
        }
      }
      size+=1
    }
    print_Writer.close()
    val duration = (System.nanoTime - t1) / 1e9d
    print("Duration: ")
    print(duration)
  }
}

File: task2.py
#command to run the code
#python second2.py 20 50 ta-feng-grocery-dataset\ta_feng_all_months_merged.csv second.txt
from pyspark import SparkContext, SparkConf
import pyspark
import sys
import time
import math
import itertools
import csv

start_time=time.time()

#creating a spark context
sc = SparkContext('local[*]','second')
sc.setLogLevel('ERROR')

#take command line inputs
filter_threshold=int(sys.argv[1])
threshold=int(sys.argv[2])
input_path = sys.argv[3]
output_path = sys.argv[4]

#reading the dataset into an RDD
largeRDD=sc.textFile(input_path).map(lambda line: line.split(',')).filter(lambda line: True if(line[0]!='"TRANSACTION_DT"') else False).map(lambda line: (line[0][1:-1]+"-"+line[1][1:-1],[line[5][1:-1].lstrip("0")])).persist(pyspark.StorageLevel.DISK_ONLY)
intermediateRDD=largeRDD.collect()

with open('intermediate.csv', 'w') as f:
    writer = csv.writer(f , lineterminator='\n')
    for tup in intermediateRDD:
        writer.writerow(tup)

#making user/business baskets
basketsRDD=largeRDD.reduceByKey(lambda a,b: a+b).filter(lambda line: True if len(line[1])>filter_threshold else False).persist(pyspark.StorageLevel.DISK_ONLY)

#toatl number of baskets
total_baskets=basketsRDD.count()

def apriori(iterator):
    baskets=[]
    baskets_in_partition=0
    itemset_size=1
    candidate=set()
    frequent_itemset=set()
    
    #get all the baskets
    for x in iterator:
        baskets.append(x[1])
    #get candidate frequent items of size one
    for basket in baskets:
        for item in basket:
            frequent_itemset.add((item,))
    
    frequent_itemset=list(frequent_itemset)
    #print(frequent_itemset)
    baskets_in_partition=len(baskets)
    reduced_threshold=threshold*(baskets_in_partition/total_baskets)
    reduced_threshold=math.floor(reduced_threshold)
    while len(frequent_itemset)!=0:  
        #print(frequent_itemset) 
        frequency=dict() 
        for basket in baskets: 
            basket=set(basket)
            for tpl in frequent_itemset:
                flag=1
                for item in tpl:
                    if item not in basket:
                        flag=0
                if flag==1:
                    if tpl in frequency:
                        frequency[tpl]+=1
                    else:
                        frequency[tpl]=1
        frequent_itemset.clear()
        for i in frequency:
            if frequency[i]>=reduced_threshold:
                candidate.add((i,1))
                frequent_itemset.append(i)
        
        frequent_itemset=set([item for t in frequent_itemset for item in t])
        itemset_size+=1
        frequent_itemset=list(itertools.combinations(frequent_itemset,itemset_size))
    
        temp=[]
        for tpl in frequent_itemset:
            temp1=list(tpl)
            temp1.sort()
            temp1=tuple(temp1)
            temp.append(temp1)
        frequent_itemset=temp
        #print(frequent_itemsets)
        #print(candidate)
        #print("************************************")
    return candidate

candidates=basketsRDD.mapPartitions(apriori).reduceByKey(lambda a,b: 1).map(lambda line: line[0]).collect()
candidates.sort()

can_output=dict()
for tpl in candidates:
    if len(tpl) in can_output:
        can_output[len(tpl)]=can_output[len(tpl)]+[tpl]
    else:
        can_output[len(tpl)]=[tpl]

print("candidates are:")
print(can_output)
print("__________________________________________________")

#definining function for findng the actual frequent pairs
def apriori_1(iterator):
    frequency=dict()
    frequent_itemset=set()
    for x in iterator:
        basket=set(x[1])
        for tpl in candidates:
            flag=1
            for item in tpl:
                if item not in basket:
                    flag=0
            if flag==1:
                if tpl in frequency:
                    frequency[tpl]+=1
                else:
                    frequency[tpl]=1
    for i in frequency:
        frequent_itemset.add((i,frequency[i]))
    #print(frequent_itemset)
    #print("************************************")
    return frequent_itemset

frequent_itemsets=basketsRDD.mapPartitions(apriori_1).reduceByKey(lambda a,b:a+b).filter(lambda line: True if line[1]>=threshold else False).map(lambda line: line[0]).collect()
frequent_itemsets.sort()
final_output=dict()
for tpl in frequent_itemsets:
    if len(tpl) in final_output:
        final_output[len(tpl)]=final_output[len(tpl)]+[tpl]
    else:
        final_output[len(tpl)]=[tpl]
print("frequent itemsets are:")
print(final_output)
print("__________________________________________________")

out_file = open(output_path,"w") 
out_file.write("Candidates:\n")
for size in can_output:
    if size==1:
        for item in can_output[size][:-1]:
            temp=str(item).replace(",","")
            out_file.write(temp+",")
        temp=str(can_output[size][-1]).replace(",","")
        out_file.write(temp)
    else:
        for item in can_output[size][:-1]:
            out_file.write(str(item)+",")
        out_file.write(str(can_output[size][-1]))
    out_file.write("\n\n")
out_file.write("Frequent Itemsets:\n")
for size in final_output:
    if size==1:
        for item in final_output[size][:-1]:
            temp=str(item).replace(",","")
            out_file.write(temp+",")
        temp=str(final_output[size][-1]).replace(",","")
        out_file.write(temp)
    else:
        for item in final_output[size][:-1]:
            out_file.write(str(item)+",")
        out_file.write(str(final_output[size][-1]))
    if(size!=len(final_output)):
        out_file.write("\n\n")

out_file.close() 
print("Duration: %s" % (time.time() - start_time))
